src/python/pants/backend/jvm/tasks/jvm_compile/rsc/rsc_compile.py

```diff
@@ -0,0 +1,603 @@
+# coding=utf-8
+# Copyright 2018 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+from __future__ import absolute_import, division, print_function, unicode_literals
+
+import functools
+import json
+import logging
+import os
+import re
+
+from six import text_type
+
+from pants.backend.jvm.subsystems.dependency_context import DependencyContext  # noqa
+from pants.backend.jvm.subsystems.jvm_platform import JvmPlatform
+from pants.backend.jvm.subsystems.shader import Shader
+from pants.backend.jvm.tasks.jvm_compile.compile_context import CompileContext
+from pants.backend.jvm.tasks.jvm_compile.execution_graph import Job
+from pants.backend.jvm.tasks.jvm_compile.zinc.zinc_compile import BaseZincCompile, ZincCompile
+from pants.base.build_environment import get_buildroot
+from pants.base.exceptions import TaskError
+from pants.base.workunit import WorkUnitLabel
+from pants.engine.fs import DirectoryToMaterialize, PathGlobs, PathGlobsAndRoot
+from pants.engine.isolated_process import ExecuteProcessRequest, FallibleExecuteProcessResult
+from pants.java.jar.jar_dependency import JarDependency
+from pants.util.contextutil import Timer
+from pants.util.dirutil import fast_relpath, safe_delete, safe_mkdir, safe_rmtree
+
+
+#
+# This is a subclass of zinc compile that uses both Rsc and Zinc to do
+# compilation.
+# It uses Rsc and the associated tools to outline scala targets. It then
+# passes those outlines to zinc to produce the final compile artifacts.
+#
+#
+logger = logging.getLogger(__name__)
+
+
+def relpathify(collection):
+  # Converts a list of paths into relpath'd paths.
+  buildroot = get_buildroot()
+  def relpath(p):
+    if p[0] != '/':
+      raise ValueError('Unexpected non-absolute path {}'.format(p))
+    return fast_relpath(p, buildroot)
+  return [relpath(c) for c in collection]
+
+
+def relpathalone(path):
+  # Converts a single path to a relpath
+  return relpathify([path])[0]
+
+
+def stdout_contents(wu):
+  if isinstance(wu, FallibleExecuteProcessResult):
+    return wu.stdout.rstrip()
+  with open(wu.output_paths()['stdout']) as f:
+    return f.read().rstrip()
+
+
+class RscCompileContext(CompileContext):
+  def __init__(self,
+               target,
+               analysis_file,
+               classes_dir,
+               rsc_mjar_file,
+               jar_file,
+               log_dir,
+               zinc_args_file,
+               sources,
+               rsc_outline_dir,
+               outline_dir):
+    super(RscCompileContext, self).__init__(target, analysis_file, classes_dir, jar_file,
+                                               log_dir, zinc_args_file, sources)
+    self.rsc_mjar_file = rsc_mjar_file
+    self.rsc_outline_dir = rsc_outline_dir
+    self.outline_dir = outline_dir
```

Comment:
I'd like propose the following directory structure of Rsc-based artifacts and naming scheme for those artifacts:

* ... referred to as `target_workdir` (in this PR: the same)
  * `rsc_compile` (in this PR: `mjar`)
    * `index` referred to as `rsc_index_dir` (in this PR: `semanticdb` aka `rsc_outline_dir`)
    * `outline` referred to as `rsc_outline_dir` (in this PR: `rsc` aka `outline_dir`)
    * `mjar.jar` referred to as `rsc_mjar_file` (in this PR: the same)
  * `zinc_compile`
    * ...

Reply:
So, like this?
```
rsc_compile/
  - index/ 
       semanticdbs from metacp/metai
  - outline/
       semanticdbs from running rsc/metai
  mjar.jar -- the mjar produced from the outlines
zinc_compile/
  <usual zinc compile elements>
```

Reply:
Yes. Thank you!